src/docs/binding-generics.md

```diff
@@ -236,3 +236,194 @@ The proposed API will only work with types known at compile time. In some cases
+
+If we consider the four cases:
+
+1. Non-frozen struct and at least one non-frozen generic parameter
+2. Non-frozen struct and (recursively) frozen generic parameters
+3. Frozen struct and at least one non-frozen generic parameter
+4. Frozen struct and (recursively) frozen generic parameters
+
+The cases one to three are simple from the projection perspective. The instantiated generic will be passed using indirection (as a non-frozen struct would in previous examples).
+
+The case four is problematic, as instances of a closed generic can be passed using swift lowering algorithm. Lets look into this more detail:
+
+```swift
+@frozen
+public struct Pair<T, U> {...}
+
+@frozen
+public struct FrozenStruct {
+    let a: Double
+    let b : Double
+}
+
+public struct NonFrozenStruct {
+    let a: Double
+    let b : Double
+}
+
+// Case 3
+public func withAtLeastOneNonFrozenParameter(data: Pair<FrozenStruct, NonFrozenStruct>) -> Pair<FrozenStruct, NonFrozenStruct> {
+    return data;
+
+// Case 4
+public func withTwoFrozenParameters(data: Pair<FrozenStruct, FrozenStruct>) -> Pair<FrozenStruct, FrozenStruct> {
+    return data;
+}
+}
+```
+
+The corresponding llvm-ir signatures of the above two functions are as follows (cases one and two would generate a signature very similar to the signature of case three).
+
+```llvm
+// Case 3
+void @"output.withAtLeastOneNonFrozenParameter(data: output.Pair<output.FrozenStruct, output.NonFrozenStruct>) -> output.Pair<output.FrozenStruct, output.NonFrozenStruct>"(ptr noalias nocapture sret(%T6output13PairVyAA0D6StructVAA0cdE0VG) %0, ptr noalias nocapture dereferenceable(32) %1)
+
+// Case 4
+{ double, double, double, double } @"output.withTwoFrozenParameters(data: output.Pair<output.FrozenStruct, output.FrozenStruct>) -> output.Pair<output.FrozenStruct, output.FrozenStruct>"(double %0, double %1, double %2, double %3)
+```
+
+Case four passes the Pair using lowering algorithm.
+
+From the user perspective we would like to have a function with the following signature:
+
+```csharp
+    public static Pair<FrozenStruct, FrozenStruct> WithTwoFrozenParameters(Pair<FrozenStruct, FrozenStruct> pair)
+```
+
+Which would then ideally call the following PInvoke.
+
+```csharp
+    [UnmanagedCallConv(CallConvs = [typeof(CallConvSwift)])]
+    public static extern Pair<FrozenStruct, FrozenStruct> WithTwoFrozenParameters(Pair<FrozenStruct, FrozenStruct> pair);
+```
+
+This however would not work as `Pair<FrozenStruct, FrozenStruct>` is a class. Runtime would try to pass the value by reference (it does not even know the layout to begin with). We could solve this problem on the projection layer.
+
+We could generate a helper struct per instance of frozen closed generic type. This could look something like:
```

Comment:
What are the `frozen closed generic type`s?

Reply:
Thanks.

> We could generate a helper struct per instance of frozen closed generic type

Would that be a source generator or some other mechanism to generate the helpers during runtime? Or do we know all the `closed generic types` during compile time (compile time of bindings vs full app, this circles back to https://github.com/dotnet/runtimelab/pull/2779#issuecomment-2491470820)?

Reply:
We should know all `closed generic types` which are part of the public API at compile time. They occur in e.g. `.swiftinterface`. That said we will need to gather that information as part of projection tooling.